### ISSUE TITLE
Finite object leak in threadpool

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10745]
|<img alt="jeffsilver's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jeffsilver reported|
|-|-|
|Trac ID|trac#1134|
|Type|defect|
|Created|2005-08-10 19:48:41Z|

Attachments:

* [threadpool_1134.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/23e/23e24d90c6975b633f20329226281e642e52ab03/9567a4823abb45fea387fbc07c5e764b3c9c584b.py) (1353 bytes) - added by therve on 2006-06-19 15:20:28Z - 
* [threadpool_test_1134.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/23e/23e24d90c6975b633f20329226281e642e52ab03/9f9f5bbdda95d4d4b45ee104c6b825dc0e2470f1.py) (2215 bytes) - added by therve on 2006-10-04 14:23:25Z - 
* [threadpool_1134_2.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/23e/23e24d90c6975b633f20329226281e642e52ab03/083a6fb2140b714eb936e5d2492ee4a9a68e066e.py) (6320 bytes) - added by therve on 2006-10-27 14:40:21Z - 
* [threadpool_reflect_1134.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/23e/23e24d90c6975b633f20329226281e642e52ab03/d9d8bf970c3c9940cd27274b6f86117894d2bf4a.diff) (2908 bytes) - added by therve on 2006-11-13 11:32:20Z - 
* [threadpool_1134_3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/23e/23e24d90c6975b633f20329226281e642e52ab03/33eed709d40b24a34244c59ae4ddb54a748c2ae1.diff) (2552 bytes) - added by therve on 2006-12-22 09:07:31Z - 
* [threadpool_1134_4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/23e/23e24d90c6975b633f20329226281e642e52ab03/0e411967205fc7c7ad33bf10ad9333103e8f93e3.diff) (2190 bytes) - added by therve on 2007-01-07 18:36:27Z - 
* [better-tests.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/23e/23e24d90c6975b633f20329226281e642e52ab03/c7963ecc26d7e94e0ee5ce89255de415f03405cd.diff) (4291 bytes) - added by antoine on 2007-04-03 15:59:53Z - 
* [better-tests2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/23e/23e24d90c6975b633f20329226281e642e52ab03/da84cfbc80e1c32fbbfe35ce3357e0f098593c18.diff) (6756 bytes) - added by antoine on 2007-04-03 17:56:01Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1134 1134
type__defect defect
reporter__jeffsilver jeffsilver
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1123703321000000 1123703321000000
changetime__1179859215000000 1179859215000000
version__ 
owner__ 
cc__exarkun cc__itamarst cc__jeffsilver cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jeffsilver's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jeffsilver commented|
|-|-|

```
#!html
<pre>
When a new thread is created in the threadpool, part of its initialization is
'args', which identifies the first task to be run. This is stored in the Thread
object as _Thread__args. Even when the thread is used later for other tasks,
this member remains. This means that any objects referred to in the args
(notably the object in which a method is to be executed) are never deleted. I've
called this a 'finite' leak, because it only leaks one set of objects per thread
in the pool. It's not the sort of leak that grows and grows.
It can be fixed by inserting the marked line in threadpool.py:

    def _worker(self, o):
        ct = threading.currentThread()
=>      del ct._Thread__args    # forget the start-up args
        while 1:
            ...
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
A better solution might to be start new threads without a task, they'll just get
one from the queue.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="jeffsilver's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jeffsilver commented|
|-|-|

```
#!html
<pre>
Yes, that would be better. I wasn't very happy with changing _Thread__args
directly, as it's supposed to be private.
I wasn't sure whether the first job needed special handling. I guess, at
present, it isn't put on the queue (because the starting of the new thread runs
it), but I haven't looked closely enough to be certain.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

Attach a patch correcting the behaviour, ie not adding the arguments at the creation of the Thread. 

There was a little caveat: the check in _startSomeWorkers on len(self.waiters)/self.q.qsize didn't worked anymore... self.waiter isn't modified during the call to startAWorker, but the length of the queue was decreased. As self.threads is modified, it seemed logical to use it in the test.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve|
|-|-|

Missing test coverage

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

Added a test for presence of Thread arguments. It's not really clean because it uses _Thread!__args, but it's better in the tests than it the code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Committed fix in threadpool-leak-1134 branch.  Didn't like _Thread__args-using test, so wrote a new one which mocks the Thread class.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

I removed review because I spotted a problem with RaceConditionTestCase. There are intermittent failure (10 %), much more frequent than before. It's based on the use of qsize() which can't be trust. Maybe the tests are too strict for a thread problem, but I'll try to find a solution.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

I've been considering parameterizing the thread creating object so that tests can provide a deterministic mock implementation of Thread.  This should make it possible for the tests which cover code which actually works right now to succeed deterministically.  Fortunately, it will also make the tests for code which does _not_ work right now (RaceConditionTestCase) fail deterministically.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

Replying to [exarkun](#note_14):
> Fortunately, it will also make the tests for code which does _not_ work right now (RaceConditionTestCase) fail deterministically.

I'm not convinced (yet) that there are problems in threadpool. There are certainly some corner cases, but it's something that can be expected for code dealing with threads. I've got code that runs in production using threadpool without problems (under *nix at least).

Using a mock object is interesting (and ambitious), but I don't see how you can maintain the same code coverage.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

There is a behavior of the threadpool which I have decided to classify as a bug.  testSingleThread should be able to pass deterministically as it is currently written.

I think code coverage would _increase_ with a mocked Thread class.  Currently the tests have no way to test specific codepaths through the threadpool.  They must rely on the fact that sometimes specific timings will result in the test following one code path while other timings will follow a different code path.  Each test should be able to select a specific codepath and follow it.

It may not necessarily be the case that _all_ tests which use a real thread implementation should be removed I am not convinced either way right now).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve removed owner|
|-|-|

I managed to have something that works (patch against the branch).

The good thing: testSingleThread seems to pass deterministically.
The bad thing: testRace passes, but I cheated a little in the test initialisation.

I also added some docs, fit 80 columns, made some cleanings.

If we still think tests should use a mocked Thread class, we should open another ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @therve|
|-|-|

Applied the patch to the branch.  Removed the cheating from testRace, which didn't cause the test to fail on my machine, so I checked it in like that.  Running the tests in the branch on the buildslaves now, we'll see what happens.

Any insight on why you needed to set the size like that?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve commented|
|-|-|

It seemed it comes from a step introduced by the pool creation, that makes the synchronisation event failed. I'd said that's why the debian buildbots failed (http://twistedmatrix.com/buildbot/debian-py2.5-select/builds/258/step-trial/0 and http://twistedmatrix.com/buildbot/debian64-py2.4-select/builds/170/step-trial/0).

But it seems there's another problem with 2.3 :/ (http://twistedmatrix.com/buildbot/debian-py2.3-select/builds/1842/step-trial/0).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve removed owner|
|-|-|

A fun one. python2.3 was the only one to report the good behaviour, because target link to the threadpool object, that contains the queue and the arguments (so test_threadCreationArguments was broken). That comes from the introduction of deque in 2.4, in the Queue module specifically.

So I changed the tests, re-added adjustPoolsize(minthreads=4) which cause other failures, and added deque to reflect (with tests).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set owner to @glyph|
|-|-|

Arg.  Having to explicitly add support for deque in order for the added test to fail is pretty unfortunate.  I'm not sure what I think about that and what it suggests about the quality of that test.

I still have the feeling that passing minthreads=4 is wrong.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve commented|
|-|-|

Replying to [exarkun](#note_21):
> Arg.  Having to explicitly add support for deque in order for the added test to fail is pretty unfortunate.  I'm not sure what I think about that and what it suggests about the quality of that test.

Well, I'd just fixed the test you've written :). I would be satisfied by something that tests the value of args and kwargs instead of using objgrep. 

> I still have the feeling that passing minthreads=4 is wrong.

Past experiences with threads tell me that a little hack to make it works is a reasonable cost, but we can always try to make better.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph set owner to @therve|
|-|-|

When I've tested object leaks in the past, I've used a weakref in combination with ```gc.collect()```.  I think that makes sense for this situation, because you have an object which you want to know has gone away.

It's also possible to use ```sys.getrefcount```.  Either way is fine, really, since they both test interactions with implementation accidents in Python's GC, but if you do use refcounts it's tempting to also store the offending value in a local variable.  Avoid doing that, because the debugger will sometimes it decides it wants to make copies of ```locals()``` and mess up your counts.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve removed owner|
|-|-|

I attached a patch on the branch, with the test following your suggestion. It seems to work great.

Also, should we do something for deque (list-like objects in general) support in reflect objgrep ?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph commented|
|-|-|

I've filed a separate ticket for deque: [#2323](https://github.com/twisted/twisted/issues/2323).

I don't think it's nearly as important as this issue :)

I've merged forward to ```threadpool-leak-1134-2```.

Having looked at the branch, it looks like a relatively clean fix and a quick solution, but I just got my hands dirty by adding the ```gc.collect()``` to make sure that nothing nondeterministic is hanging on to the contents of that weakref, so I can't OK it to merge.  However, now it no longer requires objgrep, and I think that the solution is relatively cleanly implemented.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_26"></a>|@radix commented|
|-|-|

This doesn't look like it's actually in review, since it looks like glyph's last comment was a -1 on the current state of the change.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph commented|
|-|-|

Huh?

No, it was definitely +1.  I just couldn't be a reviewer since I made an additional change to the branch.

Someone please review in ```threadpool-leak-1134-2```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun set owner to @glyph|
|-|-|

Something needs to happen about change to test_synchronization.  It either needs a note justifying it or (this is what I would prefer to see) to be removed and the rest of the thread code fixed.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_29"></a>|@glyph commented|
|-|-|

The ```minthreads=4``` hack is actually indicative of a serious problem in the way resource allocation is managed in the threadpool.  There's currently no correct way to write either of the "race condition" tests deterministically.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_30"></a>|@therve set owner to @therve|
|-|-|

I made another patch (against the branch) that improves the situation, and some cleanings.

Not yet errors proof though. But I think I have an idea on how to resolve this, so I'm assigning myself.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_31"></a>|@therve removed owner|
|-|-|

OK, I didn't make miracles, but the behaviour in the branch seems to be the same as in the trunk concerning intermittent failures. The bug mentioned in this ticket is resolved, which seems already good. Other work should go in [#11331](https://github.com/twisted/twisted/issues/11331) ?

Ready to review in threadpool-leak-1134-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun set owner to @therve|
|-|-|

Okay, I think the code in this branch basically looks sane now.  Thank you for pursuing this all the way to this point. :)

I have one last question.  In _startSomeWorkers, the while loop termination condition uses two variables (aside from self.max) to determine when enough threads have been created.  First, it uses the job queue size.  Second, it uses the number of actively existing worker threads.  The sum of these two gives an upper bound on the number of threads to create.  Since each active worker thread will eventually be able to process a task, it makes sense to count them towards the limit on the number of threads to make.  Since each job in the queue will need to be processed, it also makes sense to count each one of those towards that limit.  (Okay now that I've typed all that out I think I might actually understand it :P).  Here is the question: why does the code compute the size of the job queue before the loop begins, but recomputes the number of active worker threads each time through the loop?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_33"></a>|@therve commented|
|-|-|

Basically, q.qsize() can only be less during the loop, while len(self.working) might be  more (less is not interesting). But I don't think it won't change the behaviour if the test would be at the start, as generally we only start one thread.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_34"></a>|@therve commented|
|-|-|

exarkun ? Did it answer your question ?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_35"></a>|@exarkun commented|
|-|-|

Sorry, haven't had a chance to think about this again yet.  I'll try to make some time for it this weekend.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_36"></a>|@pitrou commented|
|-|-|

Uh, some of the tests look rather convoluted.

The so-called Synchronisation object is complicated for what it does. The strange dance at the beginning of the run() method is useless since this method is synchronised, and `threadable.synchronise` semantics are already tested in `test_threadable`. The `waiting` attribute could be a Queue instead of a Lock, which would allow `ThreadPoolTestCase._waitForLock` to use a timeout argument instead of busy waiting.

In `ThreadPoolTestCase._threadPool`, the method argument is unused which means that `ThreadPoolTestCase.testDispatch` and `ThreadPoolTestCase.testCallInThread` are identical.

The docstring for `RaceConditionTestCase.test_synchronisation` looks wrong. What it seems to test is that 1) the pool can launch new actions while existing actions are waiting (here on a `threading.Event`) 2) `threading.Event` works as advertised. I fail to see what "race condition" the docstring is talking about.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_37"></a>|@pitrou commented|
|-|-|

Added a patch to remove sillyness from threadpool tests.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_38"></a>|@pitrou commented|
|-|-|

The `test_singleThread` is not deterministic, because event.wait() can return before instructions following event.set() even start being executed. Thus it is fundamentally unreliable, and unfixable.

I have re-written the test using a weakref instead of a threading.Event. The weakref is attached to the function object passed to callInThread, and its callback triggers the next loop iteration. This guarantees that the next iteration takes place after the worker thread has returned to an idle state. The test is actually stricter now because it tests that the number of worker threads is never greater than 1 (rather than 2 which looked like a heuristic chosen so that the test failed rarely enough :-)).

The implementation is a bit complicated, it may be possible to make it simpler. However, the nice thing is that it never fails on my machine.

I'll attach a patch containing this new test and the test cleanups already mentioned earlier.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_39"></a>|@exarkun commented|
|-|-|

A previous patch already implemented a weakref-based test.  We decided that relying on that particular CPython implementation detail wasn't a good idea.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_40"></a>|@pitrou commented|
|-|-|

Replying to [exarkun](#note_39):
> A previous patch already implemented a weakref-based test.  We decided that relying on that particular CPython implementation detail wasn't a good idea.

I'm not sure what you call a CPython implementation detail. weakref is a standard Python module, and it seems at least supported by Jython (although I've never used Jython myself).

If it's about creating a weakref on a function object, I can create it on a dummy function argument instead.

And there is already a [changeset:19025 weakref-based test] in the present branch...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_41"></a>|@exarkun commented|
|-|-|

The part that is a CPython implementation detail is the part where the garbage collector is guaranteed to clean up the object at the right time.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_42"></a>|@pitrou commented|
|-|-|

Replying to [exarkun](#note_41):
> The part that is a CPython implementation detail is the part where the garbage collector is guaranteed to clean up the object at the right time.

Well, there is a problem if garbage collection never occurs (which is why I've had to add a gc.collect() call).
Other than that, the GC doesn't need to clean up anything "at the right time". It just has to happen at some point in the future, so that the test doesn't hang waiting. It can not occur too early, due to the way the Threadpool class is written.

Also, I'd be interested to know whether this test reliably succeeds on other platforms than mine.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_44"></a>|@therve commented|
|-|-|

Let's forget the refactor of the tests, we may open another ticket for that with antoine's patches.

The original request is fulfilled, please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_45"></a>|@exarkun set owner to @therve|
|-|-|

Looking tractable :)

  * The docstring for test_threadCreationArguments is of the form "Statement.  Refutation." which is somewhat odd.  I'd suggest "Negated statement." instead, eg:  "Test that creating threads in ... as arguments does not result in those objects never being freed."
  * The assertion about len(tp.threads) in the middle of the test doesn't seem related to the issue at hand.
  * The event.wait() call should most likely have a timeout
  * A previous version of the test also created a weakref for the worker function and made similar assertions about it as about the argument.  It'd be nice to bring this back.
  * in threadpool.py, in the _startSomeWorkers method, a better variable name than ```s``` would be nice
  * in threadpool.py, the _worker method needs a docstring

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_46"></a>|@therve commented|
|-|-|

Thanks, applied.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_48"></a>|@exarkun set owner to @therve|
|-|-|

Everything looks great and all the builders finally managed to do a completely successful build :P I'd suggest one minor change to _worker's docstring: "process" should probably be "proceed".  Please fix that then merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_49"></a>|@therve set status to `closed`|
|-|-|

(In [20385]) Merge threadpool-leak-1134-2

Author: therve
Reviewer: exarkun
Fixes [#10745](https://github.com/twisted/twisted/issues/10745)

Change the Threadpool to start new threads without a task, to prevent a finite
leak when a thread was created with an initial object.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
